napari/layers/_vectors_layer/model.py

```diff
@@ -0,0 +1,461 @@
+        # every "pixel" has 2 coordinates
+        pos = np.empty((2 * xdim * ydim, 2), dtype=np.float32)
+
+        # create coordinate spacing for x-y
+        # double the num of elements by doubling x sampling
+        xspace = np.linspace(0, stride_x*xdim, 2 * xdim, endpoint=False)
+        yspace = np.linspace(0, stride_y*ydim, ydim, endpoint=False)
+        xv, yv = np.meshgrid(xspace, yspace)
+
+        # assign coordinates (pos) to all pixels
+        pos[:, 0] = xv.flatten()
+        pos[:, 1] = yv.flatten()
+
+        # pixel midpoints are the first x-values of positions
+        midpt = np.zeros((xdim * ydim, 2), dtype=np.float32)
+        midpt[:, 0] = pos[0::2, 0]+(stride_x-1)/2
+        midpt[:, 1] = pos[0::2, 1]+(stride_y-1)/2
+
+        # rotate coordinates about midpoint to represent angle and length
+        pos[0::2, 0] = midpt[:, 0] - (stride_x / 2) * (self._length/2) * \
+                       vect.reshape((xdim*ydim, 2))[:, 0]
+        pos[0::2, 1] = midpt[:, 1] - (stride_y / 2) * (self._length/2) * \
+                       vect.reshape((xdim*ydim, 2))[:, 1]
+        pos[1::2, 0] = midpt[:, 0] + (stride_x / 2) * (self._length/2) * \
+                       vect.reshape((xdim*ydim, 2))[:, 0]
+        pos[1::2, 1] = midpt[:, 1] + (stride_y / 2) * (self._length/2) * \
+                       vect.reshape((xdim*ydim, 2))[:, 1]
+
+
+        return pos
+
+    def _convert_coords_to_coordinates(self, vect) -> np.ndarray:
+        """
+        To convert a list of coordinates of shape
+            (x-center, y-center, x-proj, y-proj) into a list of coordinates
+        Input coordinate of (N,4) becomes two output coordinates of (N,2)
+
+        :param vect: np.ndarray of shape (N, 4)
+        :return: position list of shape (2*N, 2) for vispy
+        """
+
+        # create empty vector of necessary shape
+        #   one coordinate for each endpoint of the vector
+        pos = np.empty((2 * len(vect), 2), dtype=np.float32)
+
+        # create pairs of points
+        pos[0::2, 0] = vect[:, 0]
+        pos[1::2, 0] = vect[:, 0]
+        pos[0::2, 1] = vect[:, 1]
+        pos[1::2, 1] = vect[:, 1]
+
+        # adjust second of each pair according to x-y projection
+        pos[1::2, 0] += vect[:, 2]
+        pos[1::2, 1] += vect[:, 3]
+
+        return pos
+
+    @property
+    def averaging(self) -> str:
+        """
+        Set the kernel over which to average
+        :return: string of averaging kernel size
+        """
+        return self._averaging
+    
+    @averaging.setter
+    def averaging(self, value: int):
+        """
+        Calculates an average vector over a kernel
+        :param averaging: one of "_avg_dims" above
+        :return:
+        """
+        self._averaging = value
+        self._kernel = value
+
+        # emit signal back to qt_properties for averaging
+        self.events.emit_avg()
+
+        self._refresh()
+
+    def averaging_bind_to(self, callback):
```

Comment:
Do we really need to provide this binding here? We don't do it for our other properties in our other layers, and it is not something that this layer uses by default. I think there will be a bigger conversation about how do we trigger plugin code based on events that happen in the layer (like when a shape is clicked on run a computation) but we havn't started that yet, and I'd rather merge this now without these bindings and solve these issues then

Reply:
Maybe up for discussion as I do see this as very useful to me.  It might be useful to the larger community?  It enables someone who defines "averaging" of vectors differently to implement their own calculation.  We could make it private if that helps?

Reply:
Why don't you just overwrite the averaging method directly, i.e.

```python
def myavgmethod():
    print('new method')

vect._default_avg = myavgmethod
```

to me this is much cleaner than having the `averaging_bind_to` function and something that only fairly advanced users will want to do. I just gave it a try and it works fine

Reply:
ah yes... this makes sense and is embarrassingly obvious, especially if we would make the above method private anyway.  Will do.